### PR TITLE
Fixes #2386. Add more super-bounded types tests

### LINKED ISSUE #2386: Add more tests for super-bounded types
We have no tests tor the following statement

> It is a compile-time error if a parameterized type T is super-bounded when
> it is used in any of the following ways:
> • T is an immediate subterm of a new expression (17.13.1) or a constant
> object expression (17.13.2).
> • T is an immediate subterm of a redirecting factory constructor signature
> (10.7.2).
> • T is an immediate subterm of an extends clause of a class (10.9), or it
> occurs as an element in the type list of an implements clause (10.10), or
> a with clause (10).

We also have no any test checking that `class C<T extends T> {}` is an error

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: Language/Generics/Superbounded_types/well_bounded_A01_t04.dart [hunk 1]
```diff
@@ -16,10 +16,17 @@ class B1<X extends A<Object?>> {}
 class B2<X extends A<dynamic>> {}
 class B3<X extends A<void>> {}
 class B4<X extends A<Never>> {}
```
Comment (eernstg) [inline Language/Generics/Superbounded_types/well_bounded_A01_t04.dart:18]: This one is actually not super-bounded.

File: Language/Generics/Superbounded_types/well_bounded_A01_t04.dart [hunk 2]
```diff
@@ -16,10 +16,17 @@ class B1<X extends A<Object?>> {}
...
+class B5<X extends A<X>> {}
```
Comment (eernstg) [inline Language/Generics/Superbounded_types/well_bounded_A01_t04.dart:19]: Ditto.

File: Language/Generics/Superbounded_types/well_bounded_A01_t04.dart [hunk 3]
```diff
@@ -16,10 +16,17 @@ class B1<X extends A<Object?>> {}
...
+
+mixin M<X extends A<X>> on A<X> {}
```
Comment (eernstg) [inline Language/Generics/Superbounded_types/well_bounded_A01_t04.dart:21]: Same here, but in order to cover mixins we could do this:
```suggestion
mixin M<X extends A<Object?>> on A<Never> {}
```

File: Language/Generics/Superbounded_types/well_bounded_A01_t04.dart [hunk 4]
```diff
@@ -16,10 +16,17 @@ class B1<X extends A<Object?>> {}
...
+extension type ET<X extends A<X>>(A<X> _) implements A<X> {}
```
Comment (eernstg) [inline Language/Generics/Superbounded_types/well_bounded_A01_t04.dart:22]: Something like this should suffice:
```dart
extension type E<X extends A<Object?>>(String _) implements Object {}
```
We are just interested in the bound of `X`, so there's no need to fight with the representation type or the implemented types, just use something that works.

**Review by eernstg — COMMENTED** [review]
Looks good! Just one test has a couple of cases that aren't super-bounded as claimed in the description.

**Comment by sgrekhov** [comment]
Thank you! Updated. PTAL.

**Review by eernstg — APPROVED** [review]
LGTM

**Pull request merged**